python/primihub/examples/hetero_xgb.py

```diff
@@ -0,0 +1,2007 @@
+    dataset_map = ph.context.Context.dataset_map
+
+    logger.debug("dataset_map {}".format(dataset_map))
+
+    logger.debug("role_nodeid_map {}".format(role_node_map))
+
+    logger.debug("node_addr_map {}".format(node_addr_map))
+
+    data_key = list(dataset_map.keys())[0]
+
+    eva_type = ph.context.Context.params_map.get("taskType", None)
+    if eva_type is None:
+        logger.warn(
+            "taskType is not specified, set to default value 'regression'.")
+        eva_type = "regression"
+
+    eva_type = eva_type.lower()
+    if eva_type != "classification" and eva_type != "regression":
+        logger.error(
+            "Invalid value of taskType, possible value is 'regression', 'classification'."
+        )
+        return
+
+    logger.info("Current task type is {}.".format(eva_type))
+
+    # 读取注册数据
+    data = ph.dataset.read(dataset_key=data_key).df_data
+    # data = ph.dataset.read(dataset_key='train_hetero_xgb_host').df_data
+    # data = pd.read_csv(
+    #     '/primihub/data/FL/hetero_xgb/train/epsilon_normalized.t.host',
+    #     header=0)
+    # data = data.iloc[:, 550:]
+
+    # y = data.pop('Class').values
+
+    print("host data: ", data)
+
+    if len(role_node_map["host"]) != 1:
+        logger.error("Current node of host party: {}".format(
+            role_node_map["host"]))
+        logger.error("In hetero XGB, only dataset of host party has label, "
+                     "so host party must have one, make sure it.")
+        return
+
+    host_nodes = role_node_map["host"]
+    host_port = node_addr_map[host_nodes[0]].split(":")[1]
+
+    guest_nodes = role_node_map["guest"]
+    guest_ip, guest_port = node_addr_map[guest_nodes[0]].split(":")
+
+    proxy_server = ServerChannelProxy(host_port)
+    proxy_server.StartRecvLoop()
+
+    proxy_client_guest = ClientChannelProxy(guest_ip, guest_port, "guest")
+
+    if 'id' in data.columns:
+        data.pop('id')
+
+    Y = data.pop('y').values
+    X_host = data.copy()
+
+    lookup_table_sum = {}
+
+    # if is_encrypted:
+    xgb_host = XGB_HOST_EN(n_estimators=num_tree,
+                           max_depth=max_depth,
+                           reg_lambda=1,
+                           sid=0,
+                           min_child_weight=min_child_weight,
+                           objective='logistic',
+                           proxy_server=proxy_server,
+                           proxy_client_guest=proxy_client_guest,
+                           encrypted=is_encrypted)
+    # channel.recv()
+    # xgb_host.channel.send(xgb_host.pub)
+    proxy_client_guest.Remote(xgb_host.pub, "xgb_pub")
+    # proxy_client_guest.Remote(public_k, "xgb_pub")
+    # print(xgb_host.channel.recv())
+    # y_hat = np.array([0.5] * Y.shape[0])
+    # ray.init()
+    # pai_actor = PaillierActor(xgb_host.prv, xgb_host.pub)
+    paillier_encryptor = ActorPool(
+        [PaillierActor.remote(xgb_host.prv, xgb_host.pub) for _ in range(10)])
```

Comment:
Does 10 actor is enough in multiple core cpu? such as our node1 server, it has 72 logic core.

Reply:
A good question. Interestingly, I haven't acquire a distinct efficiency performance set 10 or 20 in the machine with 20 kernels. And more actors also consumes more constructing time. Maybe, it's necessary to take a adequate comparison for ray parallel mechanism。